# Aqui trabajamos con https://archive.ics.uci.edu/dataset/45/heart+disease  

In [2]:
import pandas as pd

import numpy as np

import matplotlib.pyplot as plt

import matplotlib as plt

import seaborn as sns

Descaragamos el dataset

In [3]:
pip install ucimlrepo

In [4]:
from ucimlrepo import fetch_ucirepo

# fetch dataset
heart_disease = fetch_ucirepo(id=45)

# data (as pandas dataframes)
X = heart_disease.data.features
y = heart_disease.data.targets

# metadata
print(heart_disease.metadata)

# variable information
print(heart_disease.variables)


{'uci_id': 45, 'name': 'Heart Disease', 'repository_url': 'https://archive.ics.uci.edu/dataset/45/heart+disease', 'data_url': 'https://archive.ics.uci.edu/static/public/45/data.csv', 'abstract': '4 databases: Cleveland, Hungary, Switzerland, and the VA Long Beach', 'area': 'Health and Medicine', 'tasks': ['Classification'], 'characteristics': ['Multivariate'], 'num_instances': 303, 'num_features': 13, 'feature_types': ['Categorical', 'Integer', 'Real'], 'demographics': ['Age', 'Sex'], 'target_col': ['num'], 'index_col': None, 'has_missing_values': 'yes', 'missing_values_symbol': 'NaN', 'year_of_dataset_creation': 1989, 'last_updated': 'Fri Nov 03 2023', 'dataset_doi': '10.24432/C52P4X', 'creators': ['Andras Janosi', 'William Steinbrunn', 'Matthias Pfisterer', 'Robert Detrano'], 'intro_paper': {'ID': 231, 'type': 'NATIVE', 'title': 'International application of a new probability algorithm for the diagnosis of coronary artery disease.', 'authors': 'R. Detrano, A. Jánosi, W. Steinbrunn, M

In [5]:
print(heart_disease.metadata['name'])

Heart Disease


In [6]:
import requests

# Get the direct download URL from the metadata
data_url = heart_disease.metadata['data_url']

# Download the CSV directly into a pandas DataFrame
df_heart = pd.read_csv(data_url)

# Save the DataFrame to a local CSV file
df_heart.to_csv('heart_disease.csv', index=False)

print(f"Dataset downloaded and saved as 'heart_disease.csv' from: {data_url}")

Dataset downloaded and saved as 'heart_disease.csv' from: https://archive.ics.uci.edu/static/public/45/data.csv


In [7]:
# Display the first 5 rows of the downloaded dataframe
display(df_heart.head())

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,63,1,1,145,233,1,2,150,0,2.3,3,0.0,6.0,0
1,67,1,4,160,286,0,2,108,1,1.5,2,3.0,3.0,2
2,67,1,4,120,229,0,2,129,1,2.6,2,2.0,7.0,1
3,37,1,3,130,250,0,0,187,0,3.5,3,0.0,3.0,0
4,41,0,2,130,204,0,2,172,0,1.4,1,0.0,3.0,0


In [8]:
# cargar la base de datos
Heart2 = pd.read_csv("heart_disease.csv")

Vemos el tamanio de la dataset

In [9]:
Heart2.shape

(303, 14)

In [13]:
Heart2.isnull().sum()

,0
age,0
sex,0
cp,0
trestbps,0
chol,0
fbs,0
restecg,0
thalach,0
exang,0
oldpeak,0


In [14]:
Heart2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        299 non-null    float64
 12  thal      301 non-null    float64
 13  num       303 non-null    int64  
dtypes: float64(3), int64(11)
memory usage: 33.3 KB


En este caso para la variable ca, vamos a llenar los valores faltantes con la mediana

In [17]:
# Calcular la mediana de la columna 'ca'
median_ca = Heart2["ca"].median()

# Rellenar los valores nulos con la mediana
Heart2["ca"] = Heart2["ca"].fillna(median_ca)

# Verificar que ya no existan valores nulos
print(Heart2["ca"].isnull().sum())

0


In [18]:
# Imputar con el modo (valor más frecuente)
mode_thal = Heart2["thal"].mode()[0]
Heart2["thal"] = Heart2["thal"].fillna(mode_thal)

# Crear columna indicadora de nulos originales
Heart2["thal_missing"] = Heart2["thal"].isna().astype(int)

# Verificar
print(Heart2["thal"].isnull().sum())
print(Heart2["thal_missing"].value_counts())

0
thal_missing
0    303
Name: count, dtype: int64


Veamos si hay duplicados

In [19]:
# Contar cuántas filas duplicadas hay
print("Número de duplicados:", Heart2.duplicated().sum())

Número de duplicados: 0


In [20]:
# Guardar Heart2 en un archivo CSV
Heart2.to_csv("Heart2.csv", index=False)

print("Archivo Heart2.csv descargado correctamente")

Archivo Heart2.csv descargado correctamente


| Columna    | Descripción                                                                                               |
|:-----------|:----------------------------------------------------------------------------------------------------------|
| `age`      | Edad en años.                                                                                             |
| `sex`      | Sexo (1 = masculino; 0 = femenino).                                                                       |
| `cp`       | Tipo de dolor de pecho: 1: angina típica, 2: angina atípica, 3: dolor no anginoso, 4: asintomático.       |
| `trestbps` | Presión arterial en reposo (en mm Hg al ingreso hospitalario).                                            |
| `chol`     | Colesterol sérico en mg/dl.                                                                               |
| `fbs`      | Azúcar en sangre en ayunas > 120 mg/dl (1 = verdadero; 0 = falso).                                        |
| `restecg`  | Resultados del electrocardiograma en reposo: 0: normal, 1: anomalía de la onda ST-T, 2: hipertrofia ventricular izquierda probable o definida. |
| `thalach`  | Frecuencia cardíaca máxima alcanzada.                                                                      |
| `exang`    | Angina inducida por el ejercicio (1 = sí; 0 = no).                                                        |
| `oldpeak`  | Depresión del ST inducida por el ejercicio en relación con el reposo.                                     |
| `slope`    | La pendiente del segmento ST del pico de ejercicio: 1: ascendente, 2: plano, 3: descendente.              |
| `ca`       | Número de vasos principales (0-3) coloreados por fluoroscopia.                                            |
| `thal`     | Resultado de la prueba de estrés con talio: 3: normal, 6: defecto fijo, 7: defecto reversible.            |
| `num`      | Diagnóstico de enfermedad cardíaca (valor de 0 a 4, donde >0 indica enfermedad).                         |